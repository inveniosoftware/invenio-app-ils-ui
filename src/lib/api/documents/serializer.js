import { recordResponseSerializer } from '@api/utils';

const DocumentSerializers = {
  responseSerializer: function(hit) {
    return recordResponseSerializer(hit);
  },
};

export const documentSerializer = {
  fromJSON: DocumentSerializers.responseSerializer,
};
