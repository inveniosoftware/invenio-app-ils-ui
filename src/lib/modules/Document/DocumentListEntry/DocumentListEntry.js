import DocumentAuthors from '@modules/Document/DocumentAuthors';
import DocumentLanguages from '@modules/Document/DocumentLanguages';
import LiteratureCover from '@modules/Literature/LiteratureCover';
import LiteratureTags from '@modules/Literature/LiteratureTags';
import LiteratureTitle from '@modules/Literature/LiteratureTitle';
import { FrontSiteRoutes } from '@routes/urls';
import _get from 'lodash/get';
import _isEmpty from 'lodash/isEmpty';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import Truncate from 'react-truncate';
import { Grid, Item, Label, List } from 'semantic-ui-react';

export default class DocumentListEntry extends Component {
  constructor(props) {
    super(props);
    this.metadata = props.metadata;
  }

  renderCirculationInfo = meta => {
    if (
      meta.circulation &&
      (meta.circulation.available_items_for_loan_count > 0 ||
        meta.eitems.total > 0)
    ) {
      return (
        <List className="document-availability-wrapper">
          {meta.circulation.available_items_for_loan_count > 0 ? (
            <List.Item>
              <List.Icon color="green" name="check" />
              <List.Content>Available for loan</List.Content>
            </List.Item>
          ) : null}
          {meta.eitems.total > 0 ? (
            <List.Item>
              <List.Icon name="desktop" />
              <List.Content>Electronic version</List.Content>
            </List.Item>
          ) : null}
        </List>
      );
    }
    return null;
  };

  renderImprintInfo = () => {
    if (!_isEmpty(this.metadata.imprint)) {
      return (
        <Grid.Column width={4}>
          <List>
            <List.Item>
              <List.Content>
                <span>Published by </span>
                {this.metadata.imprint.publisher}
              </List.Content>
            </List.Item>
            <List.Item>
              {this.metadata.imprint.place}, {this.metadata.imprint.date}
            </List.Item>
          </List>
        </Grid.Column>
      );
    }
    return null;
  };

  renderImage = () => {
    const { volume } = this.props;
    const image = (
      <Item.Image
        floated="left"
        as={Link}
        to={FrontSiteRoutes.documentDetailsFor(this.metadata.pid)}
      >
        <LiteratureCover
          isRestricted={_get(this, 'metadata.restricted', false)}
          size="small"
          url={_get(this, 'metadata.cover_metadata.urls.medium')}
        />
      </Item.Image>
    );

    if (volume) {
      return (
        <div className="search-result-image">
          <Label floating color="black">
            Volume {volume}
          </Label>
          {image}
        </div>
      );
    }

    return image;
  };

  render() {
    return (
      <Item>
        {this.renderImage()}
        <Item.Content>
          <Item.Meta>{this.metadata.document_type}</Item.Meta>
          <Item.Header
            as={Link}
            to={FrontSiteRoutes.documentDetailsFor(this.metadata.pid)}
          >
            <LiteratureTitle title={this.metadata.title} />
          </Item.Header>
          <Item.Meta>
            <DocumentAuthors
              authors={this.metadata.authors}
              hasOtherAuthors={this.metadata.other_authors}
              prefix="by "
              limit={10}
            />
          </Item.Meta>
          <Item.Description>
            <Truncate lines={2} ellipsis="... ">
              {this.metadata.abstract}
            </Truncate>
          </Item.Description>
          <Item.Meta>
            <Grid>
              <Grid.Column width={4}>
                {this.renderCirculationInfo(this.metadata)}
              </Grid.Column>
              {this.renderImprintInfo()}
              <Grid.Column width={4}>
                <List>
                  {this.metadata.edition && (
                    <List.Item>
                      <List.Content>
                        <span>Edition: </span>
                        {this.metadata.edition}
                      </List.Content>
                    </List.Item>
                  )}
                  {this.metadata.languages && (
                    <List.Item>
                      <List.Content>
                        <span>Languages: </span>
                        <DocumentLanguages
                          languages={this.metadata.languages}
                        />
                      </List.Content>
                    </List.Item>
                  )}
                </List>
              </Grid.Column>
            </Grid>
          </Item.Meta>
          <Item.Extra>
            <LiteratureTags tags={this.metadata.tags} />
          </Item.Extra>
        </Item.Content>
      </Item>
    );
  }
}

DocumentListEntry.propTypes = {
  metadata: PropTypes.object.isRequired,
  volume: PropTypes.string,
};

DocumentListEntry.defaultProps = {
  volume: null,
};
