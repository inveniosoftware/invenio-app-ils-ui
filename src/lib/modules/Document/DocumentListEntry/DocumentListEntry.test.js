import * as testData from '@testData/documents.json';
import { shallow } from 'enzyme';
import React from 'react';
import DocumentListEntry from './DocumentListEntry';

it('should render correctly', () => {
  const data = {
    metadata: {
      ...testData[0],
      pid: '13',
      eitems: { hits: [], total: 0 },
      circulation: {
        available_items_for_loan_count: 0,
      },
    },
  };

  const component = shallow(<DocumentListEntry metadata={data.metadata} />);
  expect(component).toMatchSnapshot();
});
