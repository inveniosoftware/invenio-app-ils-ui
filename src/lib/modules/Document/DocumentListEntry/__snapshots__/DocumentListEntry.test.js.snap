// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentListEntry
  metadata={
    Object {
      "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
      "abstract": "This is an abstract",
      "authors": Array [
        Object {
          "full_name": "Jack E. Davis",
        },
      ],
      "circulation": Object {
        "available_items_for_loan_count": 0,
      },
      "document_type": "BOOK",
      "edition": "1",
      "eitems": Object {
        "hits": Array [],
        "total": 0,
      },
      "keywords": Object {
        "source": "X",
        "value": "Patata",
      },
      "languages": Array [
        "en",
      ],
      "pid": "13",
      "publication_year": "1950",
      "title": "The Gulf: The Making of An American Sea",
    }
  }
/>
`;
